### PR TITLE
New plugin for powerd support

### PR DESCRIPTION
**Pull request by g2pop**
The fu-plugin-upower.c was copied and renamed to fu-plugin-powerd.c to
fit the project it will be used for. This CL then rids the fu-plugin-powerd.c
file of unecessary functions and variables and deleting whitespace. This
leaves the file with just enough functions to be enabled on fwupd and
permit all updates.

The upower README.md file was re-created to exclude information specific to upower
and include information on powerd and the specific interface it will
use.
The meson.build file specific to upower was also re-created to be powerd-specific and
call fu-plugin-powerd.c. Also, the .conf install function was removed to be consistent with
an upstream change as it was deemed unecessary.

The meson.build general to all fwupd plugins was edited to include the
powerd subdirectory as well, so that the plugin will be enabled upon
compilation.

The .quirk file has been removed since updates to fwupd will create
a general daemon.quirk file for upower and powerd to share.

Change-Id: I7542b02d21a22248337fad966874c64cf495c0a2

Type of pull request:
- [x] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [ ] Feature
- [ ] Documentation

### CODE REVIEW

**Comment by bleungatchromium** [comment]
@binaryNewts @djcampello as fyi.

**Review by bleungatchromium — COMMENTED** [review]
Please remove the "Change-Id:" in the commit message. (looks like an artifact of the chromiumos environment you all are using).

File: plugins/powerd/fu-plugin-powerd.c
Comment (hughsie) [inline plugins/powerd/fu-plugin-powerd.c:32]: not two-space tabs, one 8-char tab for indent please.

File: plugins/powerd/meson.build
Comment (hughsie) [inline plugins/powerd/meson.build:4]: do we want to build this on all systems, e.g. Fedora, Ubuntu, Arch etc? isn't powerd really ChromeOS-specific? If that's true, maybe we need an option in meson_options.txt which gets enabled in ChromeOS (and possibly one of the CI targets for testing). If it's just a D-Bus service which fails on startup then I guess it's fine to leave this compiled on all distros if it makes it easier to test things.

File: plugins/powerd/fu-plugin-powerd.c
Comment (hughsie) [inline plugins/powerd/fu-plugin-powerd.c:3]: You need to follow the existing style, e.g. this should be

     * Copyright (C) 2021 Twain Byrnes <binaryNewts@google.com>
     * Copyright (C) 2021 George Popoola <gpopoola@google.com>

**Comment by hughsie** [comment]
@g2pop feel free to `git push --force` to this branch, and the PR will update. It's also fine to push commits with a message of "FIXUP" as you develop the plugin -- we can squash all the commits down into a msingle sensible and beautiful one when we're ready to merge.

File: plugins/powerd/meson.build
Comment (superm1) [reply plugins/powerd/meson.build:4]: I was thinking about this, and I think ideally what we want is a meson option like `plugin_power` that can be set to `none`, `upower`, `powerd`.  There is never going to be a situation that you want both plugins running in Fedora for example.  It's going to be you use one or the other or none.
For the purpose of our CI we can run it in `powerd` mode for something like the Arch CI, in `none` mode for the purpose of Windows, and then `upower` for all the rest.

Comment (hughsie) [reply plugins/powerd/meson.build:4]: on the fiip side, if one of the plugins self-disable at startup it's one less configure option and one step closer to "just works". I also wonder if arch (or another CI target) can actually run `powerd` for testing.

Comment (superm1) [reply plugins/powerd/meson.build:4]: That's true.  I guess let'd hold this decision until there is more meat to the plugin.

File: plugins/powerd/fu-plugin-powerd.c
Comment (g2pop) [reply plugins/powerd/fu-plugin-powerd.c:4]: To clarify, did you mean the extra tab before "George Popoola ..." or the skipped lines before and after the copyright?

Comment (hughsie) [reply plugins/powerd/fu-plugin-powerd.c:4]: If in doubt, just make it match the existing style: e.g. https://github.com/fwupd/fwupd/blob/master/libfwupdplugin/fu-efivar.c#L2

Comment (g2pop) [reply plugins/powerd/fu-plugin-powerd.c:4]: Alright, no problem. Thank you.

**Review by bleungatchromium — COMMENTED** [review]
Hi @binaryNewts,

It looks like the last  commit you pushed to this branch contains your work in progress for some follow on work (inhibit system suspend during update).

It also isn't clean, as it looks like there's some git merge conflict artifacts inside you still need to resolve.

I would recommend pulling this out into a separate branch and leave this branch (binaryNewts:powerd-plugin-skeleton) pointing to just the work you and @g2pop have done on the skeleton step and that's it.

File: libfwupdplugin/fu-common.h
Comment (superm1) [inline libfwupdplugin/fu-common.h:74]: this looks like a bad rebase or merge

File: libfwupdplugin/fu-common.c
Comment (superm1) [inline libfwupdplugin/fu-common.c:1251]: don't change whitespace for existing stuff in your commit

File: libfwupdplugin/fu-common.h
Comment (superm1) [inline libfwupdplugin/fu-common.h:100]: looks like a bad merge/rebase

Comment (bleungatchromium) [reply libfwupdplugin/fu-common.h]: Yeah, I had @binaryNewts force push this back to a commit from earlier, and submit their Inhibit Suspend change in  separate PR, #3489. That one is going to be dependent on the initial skeleton, and I'll help them rebase that work once the skeleton is merged.

File: libfwupdplugin/fu-common.h
Comment (bleungatchromium) [reply libfwupdplugin/fu-common.h]: I had @binaryNewts force push this back to the original commit.

File: plugins/powerd/meson.build
Comment (bleungatchromium) [reply plugins/powerd/meson.build:4]: Hi @superm1 and @hughsie 

This plugin will be similar to the upower plugin in that it will gracefully disable itself on startup if the dbus interface for powerd that we're building isn't there.

So on Chrome OS, the upower plugin comes up as Disabled because it fails to connet to the Upower daemon over debus, and the same will be true on other distros for the powerd plugin.

**Comment by bleungatchromium** [comment]
@hughsie and @superm1, any other comments on this skeleton plugin submission? It needs to go in before the other work the interns are working on as well (as they've already started separate PRs).

It looks like the fedora builder failed with this error:
```
error: Installed (but unpackaged) file(s) found:
   /usr/lib/debug/usr/lib64/fwupd-plugins-3/libfu_plugin_powerd.so-1.6.2-0.1alpha.fc34.x86_64.debug
   /usr/lib64/fwupd-plugins-3/libfu_plugin_powerd.so
```

I don't remember how this was resolved before when I ran into this with mine. Anything we should be doing to get this one to go green?

**Comment by bleungatchromium** [comment]
> It looks like the fedora builder failed with this error:
> 
> ```
> error: Installed (but unpackaged) file(s) found:
>    /usr/lib/debug/usr/lib64/fwupd-plugins-3/libfu_plugin_powerd.so-1.6.2-0.1alpha.fc34.x86_64.debug
>    /usr/lib64/fwupd-plugins-3/libfu_plugin_powerd.so
> ```
> 
> I don't remember how this was resolved before when I ran into this with mine. Anything we should be doing to get this one to go green?

Hi @g2pop,

I ran into this when I added my skeleton for cros_ec as well, and this is what you need to do.
You need to add `fwupd-plugins-3/libfu_plugin_powerd.so` to the contrib/fwupd.spec.in file so that fedora knows this .so file exists and is expected.

File: plugins/powerd/fu-plugin-powerd.c
Comment (superm1) [inline plugins/powerd/fu-plugin-powerd.c:21]: in my mind this plugin should by default return FALSE and set error until there is something that it can do, or there should be more meat to it before merging anything.

Comment (bleungatchromium) [reply plugins/powerd/fu-plugin-powerd.c:21]: Yeah, agreed.

@g2pop, can you go ahead and make that change?

Comment (g2pop) [reply plugins/powerd/fu-plugin-powerd.c]: Just made the change.

Comment (hughsie) [reply plugins/powerd/fu-plugin-powerd.c]: If you're returning FALSE then you need to set the `GError` otherwise the plugin loader gets very upset. You need something like this before the return FALSE:

	g_set_error (error,
		     G_IO_ERROR,
		     G_IO_ERROR_NOT_SUPPORTED,
		     "not supported yet");

**Comment by hughsie** [comment]
Also, I think you can make the commit message 90% smaller perhaps. There's more lines in **that** than the code :)

**Review by bleungatchromium — CHANGES_REQUESTED** [review]
Hey, @g2pop, it looks like you accidentally included fu-plugin-powerd.c.orig, which looks like some merge conflict artifacts. Let's make sure to delete that from this submission.

Also, clean up the commit message again to remove the chromiumos-only BUG, TEST, and Change-Id

**Comment by bleungatchromium** [comment]
Hi @g2pop and @binaryNewts 

Please go through these two commits and make sure your editor is set to 8-wide tabs instead of spaces like the other .c/.h files in the fwupd codebase.


https://github.com/fwupd/fwupd/blob/master/CONTRIBUTING.md

File: plugins/powerd/fu-plugin-powerd.c
Comment (hughsie) [inline plugins/powerd/fu-plugin-powerd.c:82]: /* comments please */

File: plugins/powerd/fu-plugin-powerd.c
Comment (hughsie) [inline plugins/powerd/fu-plugin-powerd.c:35]: space before the {

File: plugins/powerd/fu-plugin-powerd.c
Comment (hughsie) [inline plugins/powerd/fu-plugin-powerd.c:33]: space before the bracket

File: plugins/powerd/fu-plugin-powerd.c
Comment (hughsie) [inline plugins/powerd/fu-plugin-powerd.c:15]: don't you need to unref this in destroy?

Comment (g2pop) [reply plugins/powerd/fu-plugin-powerd.c]: The fu_plugin_destoy () function was added in the latest upload.

**Comment by bleungatchromium** [comment]
Hey @g2pop,

Do you think you could apply your tab changes to the skeleton commit (Skeleton plugin for powerd) as well?

Right now, your most recent commit fixes the tabs on the skeleton, but it should be focused on the dbus changes and not style ones.

File: plugins/powerd/fu-plugin-powerd.c
Comment (bleungatchromium) [inline plugins/powerd/fu-plugin-powerd.c:69]: Something got cut off here. Please double check this!

File: plugins/powerd/fu-plugin-powerd.c
Comment (bleungatchromium) [inline plugins/powerd/fu-plugin-powerd.c:36]: Extra whitespace.

File: plugins/powerd/fu-plugin-powerd.c
Comment (bleungatchromium) [inline plugins/powerd/fu-plugin-powerd.c:73]: Looks like a duplicate {

**Review by bleungatchromium — CHANGES_REQUESTED** [review]

File: plugins/powerd/fu-plugin-powerd.c
Comment (bleungatchromium) [inline plugins/powerd/fu-plugin-powerd.c:123]: From the CI:

```
../plugins/powerd/fu-plugin-powerd.c: In function ‘fu_plugin_update_prepare’:
../plugins/powerd/fu-plugin-powerd.c:122:9: error: format ‘%u’ expects argument of type ‘unsigned int’, but argument 5 has type ‘int’ [-Werror=format=]
  122 |         "Cannot install update when system battery "
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../plugins/powerd/fu-plugin-powerd.c:123:27: note: format string is defined here
  123 |         "is not at least %u%% unless forced",
      |                          ~^
      |                           |
      |                           unsigned int
      |                          %u
cc1: all warnings being treated as errors
```

**Review by bleungatchromium — CHANGES_REQUESTED** [review]

**Review by bleungatchromium — CHANGES_REQUESTED** [review]
Commit message comment:

Once the initial commit goes in that introduces a plugin (your Skeleton), you should format the commit title as such:

`powerd: Dbus Connection and Parsing`

File: plugins/powerd/fu-plugin-powerd.c
Comment (superm1) [inline plugins/powerd/fu-plugin-powerd.c:90]: shouldn't this be prefix error?  if the powerd_response was NULL then error should have been set

File: plugins/powerd/fu-plugin-powerd.c
Comment (bleungatchromium) [inline plugins/powerd/fu-plugin-powerd.c:64]: I thought we had suggested this to return FALSE in the skeleton?

File: plugins/powerd/fu-plugin-powerd.c
Comment (bleungatchromium) [inline plugins/powerd/fu-plugin-powerd.c:91]: CI's not happy about this.  Let's keep strings to one line (don't worry about 80 char width for things like this).

```
../dist-unpack/fwupd-1.6.2/plugins/powerd/fu-plugin-powerd.c: In function ‘fu_plugin_update_prepare’:
../dist-unpack/fwupd-1.6.2/plugins/powerd/fu-plugin-powerd.c:91:21: warning: missing terminating " character
   91 |                     "Battery information from powerd was not loaded
      |                     ^
../dist-unpack/fwupd-1.6.2/plugins/powerd/fu-plugin-powerd.c:91:21: error: missing terminating " character
   91 |                     "Battery information from powerd was not loaded
      |                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../dist-unpack/fwupd-1.6.2/plugins/powerd/fu-plugin-powerd.c:92:21: error: ‘successfully’ undeclared (first use in this function)
   92 |                     successfully: ");
      |                     ^~~~~~~~~~~~
../dist-unpack/fwupd-1.6.2/plugins/powerd/fu-plugin-powerd.c:92:21: note: each undeclared identifier is reported only once for each function it appears in
../dist-unpack/fwupd-1.6.2/plugins/powerd/fu-plugin-powerd.c:92:33: error: expected ‘)’ before ‘:’ token
   92 |                     successfully: ");
      |                                 ^
      |                                 )
../dist-unpack/fwupd-1.6.2/plugins/powerd/fu-plugin-powerd.c:89:32: note: to match this ‘(’
   89 |                 g_prefix_error (
      |                                ^
../dist-unpack/fwupd-1.6.2/plugins/powerd/fu-plugin-powerd.c:92:35: warning: missing terminating " character
   92 |                     successfully: ");
      |                                   ^
../dist-unpack/fwupd-1.6.2/plugins/powerd/fu-plugin-powerd.c:92:35: error: missing terminating " character
   92 |                     successfully: ");
      |                                   ^~~
../dist-unpack/fwupd-1.6.2/plugins/powerd/fu-plugin-powerd.c:93:30: error: expected ‘;’ before ‘}’ token
   93 |                 return FALSE;
      |                              ^
      |                              ;
   94 |         }
      |         ~                   
```

Comment (bleungatchromium) [reply plugins/powerd/fu-plugin-powerd.c:91]: Actually, it looks like you were just missing a pair of "s on this line and the beginning of the next if you wanted to do it this way...

File: plugins/powerd/fu-plugin-powerd.c
Comment (g2pop) [reply plugins/powerd/fu-plugin-powerd.c:64]: Right, sorry about that, it got changed while I was resolving a merge conflict.

**Review by bleungatchromium — CHANGES_REQUESTED** [review]

File: plugins/powerd/fu-plugin-powerd.c
Comment (bleungatchromium) [inline plugins/powerd/fu-plugin-powerd.c:39]: From fwupd/CONTRIBUTING.md:

>Comments should not start with a capital letter or end with a full stop and should be C-style, not C++-style, e.g. /* this */ not // this

Comments shouldn't start with a capital, and shouldn't end with a full stop. Basically, they're not meant to be sentences.

File: plugins/powerd/fu-plugin-powerd.c
Comment (bleungatchromium) [inline plugins/powerd/fu-plugin-powerd.c:79]: Same as the other comment. Don't start a comment with a capital letter.

File: plugins/powerd/fu-plugin-powerd.c
Comment (bleungatchromium) [inline plugins/powerd/fu-plugin-powerd.c:63]: As we discussed on the call this morning, you may need to do a test method call so that systems that don't have powerd correctly disable themselves.

See how fu-plugin-upower.c does it here:
https://github.com/fwupd/fwupd/blob/master/plugins/upower/fu-plugin-upower.c#L100

**Review by bleungatchromium — CHANGES_REQUESTED** [review]

File: plugins/powerd/fu-plugin-powerd.c
Comment (bleungatchromium) [inline plugins/powerd/fu-plugin-powerd.c:96]:
```
/*
 * multi-line comments look like this
 * second line in a multi-line comment
 */
```

I also think in these cases, it's ok to keep it to one line, and ignore the 80 char limit if you go over by a few characters or one word or so.

(also, avoid capital letter to start comment)

File: plugins/powerd/fu-plugin-powerd.c
Comment (bleungatchromium) [inline plugins/powerd/fu-plugin-powerd.c:104]: Same here. No capital letter. Probably OK for this to remain on one line, don't need multi-line comment.

File: plugins/powerd/fu-plugin-powerd.c
Comment (bleungatchromium) [inline plugins/powerd/fu-plugin-powerd.c:71]: You're missing a } to close out this new block here.

File: plugins/powerd/fu-plugin-powerd.c
Comment (bleungatchromium) [inline plugins/powerd/fu-plugin-powerd.c:64]: You'll need to declare name_owner.

From CI:
```
../dist-unpack/fwupd-1.6.2/plugins/powerd/fu-plugin-powerd.c: In function ‘fu_plugin_startup’:
../dist-unpack/fwupd-1.6.2/plugins/powerd/fu-plugin-powerd.c:64:9: error: ‘name_owner’ undeclared (first use in this function)
   64 |         name_owner = g_dbus_proxy_get_name_owner (data->proxy);
      |         ^~~~~~~~~~
../dist-unpack/fwupd-1.6.2/plugins/powerd/fu-plugin-powerd.c:64:9: note: each undeclared identifier is reported only once for each function it appears in
```
Put this at the end of the declarations for _startup.
`	g_autofree gchar *name_owner = NULL;`

**Review by bleungatchromium — CHANGES_REQUESTED** [review]

File: plugins/powerd/fu-plugin-powerd.c
Comment (bleungatchromium) [inline plugins/powerd/fu-plugin-powerd.c:36]: (nit) project style has g_autoptr and g_autofree declared as the last variables in a block, so re-order this below `FuPluginData *data = ...`.

File: plugins/powerd/fu-plugin-powerd.c
Comment (bleungatchromium) [inline plugins/powerd/fu-plugin-powerd.c:84]: (nit) project style has g_autoptr and g_autofree declared as the last variables in a block, so re-order this below the other declarations.

**Review by bleungatchromium — COMMENTED** [review]
Just style nits from me right now.

**Review by bleungatchromium — APPROVED** [review]

**Pull request merged**
